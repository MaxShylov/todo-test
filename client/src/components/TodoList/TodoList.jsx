import React, { Component } from 'react';
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { arrayMove } from 'react-sortable-hoc';
import isEmpty from 'lodash/isEmpty';

import './TodoList.scss'
import SortableList from '../SortableList/SortableList';
import { getTasks, saveTasks, sortTasks } from '../../redux/actions/tasks/tasks';
import Container from '../Container/Container';
import { setStatusConnection } from '../../redux/actions/connection/connection';

const { arrayOf, object, func } = PropTypes;


export class TodoList extends Component {
  static propTypes = {
    // redux
    tasks: arrayOf(object).isRequired,
    getTasks: func.isRequired,
    sortTasks: func,
    saveTasks: func,
    setStatusConnection: func.isRequired
  };

  state = {
    tasks: this.props.tasks
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.tasks !== this.state.tasks) {
      this.setState({ tasks: nextProps.tasks });
    }
  }

  componentDidMount() {
    const { getTasks, saveTasks, setStatusConnection } = this.props;

    getTasks();
    setStatusConnection(navigator.onLine);

    window.addEventListener('beforeunload', () => {
      saveTasks();
    });

    window.addEventListener('online', () => {
      saveTasks();
      setStatusConnection(true)
    });

    window.addEventListener('offline', () => {
      setStatusConnection(false)
    });
  };


  onSortEnd = ({ oldIndex, newIndex }) => {
    const
      { tasks } = this.state,
      { sortTasks, saveTasks } = this.props,
      sortedTasks = arrayMove(tasks, oldIndex, newIndex);

    this.setState({ tasks: sortedTasks });
    sortTasks(sortedTasks);
    saveTasks()
  };

  render() {
    const { tasks } = this.state;

    return (
      <Container className="TodoList">

        <h2>Todo list:</h2>

        {isEmpty(tasks) ? 'No tasks' : (
          <SortableList
            pressDelay={150}
            tasks={tasks}
            onSortEnd={this.onSortEnd}
          />
        )}
      </Container>
    );
  }
}


const mapStateToProps = state => ({
  tasks: state.tasks.tasks
});

const matchDispatchToProps = dispatch => bindActionCreators({
  getTasks,
  sortTasks,
  saveTasks,
  setStatusConnection
}, dispatch);

export default connect(mapStateToProps, matchDispatchToProps)(TodoList);
