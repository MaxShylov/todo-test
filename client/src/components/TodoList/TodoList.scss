@import "../../styles/variables";

.TodoList {
  text-align: center;

  h2 {
    margin-top: 20px;
  }
}
