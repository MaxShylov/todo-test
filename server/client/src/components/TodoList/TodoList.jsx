import React, { Component } from 'react';
import { bindActionCreators } from 'redux'
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { arrayMove } from 'react-sortable-hoc';
import isEmpty from 'lodash/isEmpty';

import './TodoList.scss'
import SortableList from '../SortableList/SortableList';
import { getTasks, saveTasks, sortTasks } from '../../redux/actions/tasks/tasks';
import Container from '../Container/Container';
import { setStatusConnection } from '../../redux/actions/connection/connection';

const { arrayOf, object, func } = PropTypes;

let intSave = null;


export class TodoList extends Component {
  static propTypes = {
    // redux
    tasks: arrayOf(object).isRequired,
    getTasks: func.isRequired,
    sortTasks: func,
    saveTasks: func,
    setStatusConnection: func.isRequired
  };

  state = {
    tasks: this.props.tasks
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.tasks !== this.state.tasks) {
      this.setState({ tasks: nextProps.tasks });
    }
  }

  componentDidMount() {
    const { getTasks, saveTasks, setStatusConnection } = this.props;

    getTasks();
    setStatusConnection(navigator.onLine);

    if (!intSave) intSave = setInterval(saveTasks, 15000);

    window.addEventListener('beforeunload', () => {
      clearInterval(intSave);
      saveTasks();
    });

    window.addEventListener('online', () => {
      saveTasks();
      setStatusConnection(true)
    });

    window.addEventListener('offline', () => {
      setStatusConnection(false)
    });
  };

  componentWillUnmount() {
    if (intSave) clearInterval(intSave);
  }

  onSortEnd = ({ oldIndex, newIndex }) => {
    const
      { tasks } = this.state,
      { sortTasks } = this.props,
      sortedTasks = arrayMove(tasks, oldIndex, newIndex);

    this.setState({ tasks: sortedTasks });
    sortTasks(sortedTasks)
  };

  render() {
    const { tasks } = this.state;

    return (
      <Container className="TodoList">

        <h2>Todo list:</h2>

        {isEmpty(tasks) ? 'No tasks' : (
          <SortableList
            pressDelay={150}
            tasks={tasks}
            onSortEnd={this.onSortEnd}
          />
        )}
      </Container>
    );
  }
}


const mapStateToProps = state => ({
  tasks: state.tasks.tasks
});

const matchDispatchToProps = dispatch => bindActionCreators({
  getTasks,
  sortTasks,
  saveTasks,
  setStatusConnection
}, dispatch);

export default connect(mapStateToProps, matchDispatchToProps)(TodoList);
