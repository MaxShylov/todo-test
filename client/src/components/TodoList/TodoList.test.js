import React from 'react';
import { shallow } from 'enzyme'

import { TodoList } from './TodoList.jsx'


describe('TodoList', () => {

  const props = {
    tasks: [{ id: 1 }, { id: 2 }],
    getTasks: () => null,
    setStatusConnection: () => null
  };

  describe('initial', () => {
    const wrapper = shallow(<TodoList {...props} />);

    it('renders properly', () => expect(wrapper).toMatchSnapshot());
  });

});
