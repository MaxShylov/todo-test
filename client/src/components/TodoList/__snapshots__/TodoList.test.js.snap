// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList initial renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoList
    getTasks={[Function]}
    setStatusConnection={[Function]}
    tasks={
      Array [
        Object {
          "id": 1,
        },
        Object {
          "id": 2,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h2>
          Todo list:
        </h2>,
        <sortableList(SortableList)
          axis="y"
          distance={0}
          getHelperDimensions={[Function]}
          hideSortableGhost={true}
          lockOffset="50%"
          lockToContainerEdges={false}
          onSortEnd={[Function]}
          pressDelay={150}
          pressThreshold={5}
          shouldCancelStart={[Function]}
          tasks={
            Array [
              Object {
                "id": 1,
              },
              Object {
                "id": 2,
              },
            ]
          }
          transitionDuration={300}
          useWindowAsScrollContainer={false}
        />,
      ],
      "className": "TodoList",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Todo list:",
        },
        "ref": null,
        "rendered": "Todo list:",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "axis": "y",
          "distance": 0,
          "getHelperDimensions": [Function],
          "hideSortableGhost": true,
          "lockOffset": "50%",
          "lockToContainerEdges": false,
          "onSortEnd": [Function],
          "pressDelay": 150,
          "pressThreshold": 5,
          "shouldCancelStart": [Function],
          "tasks": Array [
            Object {
              "id": 1,
            },
            Object {
              "id": 2,
            },
          ],
          "transitionDuration": 300,
          "useWindowAsScrollContainer": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h2>
            Todo list:
          </h2>,
          <sortableList(SortableList)
            axis="y"
            distance={0}
            getHelperDimensions={[Function]}
            hideSortableGhost={true}
            lockOffset="50%"
            lockToContainerEdges={false}
            onSortEnd={[Function]}
            pressDelay={150}
            pressThreshold={5}
            shouldCancelStart={[Function]}
            tasks={
              Array [
                Object {
                  "id": 1,
                },
                Object {
                  "id": 2,
                },
              ]
            }
            transitionDuration={300}
            useWindowAsScrollContainer={false}
          />,
        ],
        "className": "TodoList",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Todo list:",
          },
          "ref": null,
          "rendered": "Todo list:",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "axis": "y",
            "distance": 0,
            "getHelperDimensions": [Function],
            "hideSortableGhost": true,
            "lockOffset": "50%",
            "lockToContainerEdges": false,
            "onSortEnd": [Function],
            "pressDelay": 150,
            "pressThreshold": 5,
            "shouldCancelStart": [Function],
            "tasks": Array [
              Object {
                "id": 1,
              },
              Object {
                "id": 2,
              },
            ],
            "transitionDuration": 300,
            "useWindowAsScrollContainer": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
